src/test/tests_common.h

```diff
@@ -42,7 +42,154 @@
+	int	rc;
+	int	num_ctx;
+};
+
+static inline void
+sync_timedwait(struct wfr_status *wfrs, int sec, int line_number)
+{
+	struct timespec	deadline;
+	int		rc;
+
+	rc = clock_gettime(CLOCK_REALTIME, &deadline);
+	D_ASSERTF(rc == 0, "clock_gettime() failed at line %d rc: %d\n",
+		  line_number, rc);
+
+	deadline.tv_sec += sec;
+
+	/* If this fails timeout, return and retry
+	 * else, retry based on status of rpc
+	 */
+	rc = sem_timedwait(&wfrs->sem, &deadline);
+	if (rc != 0)
+		wfrs->rc = rc;
+}
+
+static void
+ctl_client_cb(const struct crt_cb_info *info)
+{
+	struct wfr_status		*wfrs;
+	struct crt_ctl_ep_ls_out	*out_ls_args;
+	char				*addr_str;
+	int				 i;
+
+	wfrs = (struct wfr_status *)info->cci_arg;
+
+	if (info->cci_rc == 0) {
+		out_ls_args = crt_reply_get(info->cci_rpc);
+		wfrs->num_ctx = out_ls_args->cel_ctx_num;
+		wfrs->rc = out_ls_args->cel_rc;
+
+		D_DEBUG(DB_TEST, "ctx_num: %d\n",
+			out_ls_args->cel_ctx_num);
+		addr_str = out_ls_args->cel_addr_str.iov_buf;
+		for (i = 0; i < out_ls_args->cel_ctx_num; i++) {
+			D_DEBUG(DB_TEST, "    %s\n", addr_str);
+				addr_str += (strlen(addr_str) + 1);
+		}
+	} else {
+		wfrs->rc = info->cci_rc;
+	}
+
+	sem_post(&wfrs->sem);
+}
+
+int
+wait_for_ranks(crt_context_t ctx, crt_group_t *grp, d_rank_list_t *rank_list,
+		int tag, int total_ctx, double ping_timeout,
+		double total_timeout)
+{
+	struct wfr_status		ws;
+	struct timespec			t1, t2;
+	struct crt_ctl_ep_ls_in		*in_args;
+	d_rank_t			rank;
+	crt_rpc_t			*rpc = NULL;
+	crt_endpoint_t			server_ep;
+	double				time_s = 0;
+	int				i = 0;
+	int				rc = 0;
+
+	rc = d_gettime(&t1);
+	D_ASSERTF(rc == 0, "d_gettime() failed; rc=%d\n", rc);
+
+	rc = sem_init(&ws.sem, 0, 0);
+	D_ASSERTF(rc == 0, "sem_init() failed; rc=%d\n", rc);
+
+	server_ep.ep_tag = tag;
+	server_ep.ep_grp = grp;
+
+	for (i = 0; i < rank_list->rl_nr; i++) {
+
+		rank = rank_list->rl_ranks[i];
+
+		server_ep.ep_rank = rank;
+
+		rc = crt_req_create(ctx, &server_ep, CRT_OPC_CTL_LS, &rpc);
+		D_ASSERTF(rc == 0, "crt_req_create failed; rc=%d\n", rc);
+
+		in_args = crt_req_get(rpc);
+		in_args->cel_grp_id = grp->cg_grpid;
+		in_args->cel_rank = rank;
+
+		rc = crt_req_set_timeout(rpc, ping_timeout);
+		D_ASSERTF(rc == 0, "crt_req_set_timeout failed; rc=%d\n", rc);
+
+		ws.rc = 0;
+		ws.num_ctx = 0;
+
+		rc = crt_req_send(rpc, ctl_client_cb, &ws);
+
+		if (rc == 0)
+			sync_timedwait(&ws, 120, __LINE__);
+
+		while (ws.rc != 0 && time_s < total_timeout) {
+			rc = crt_req_create(ctx, &server_ep,
+					    CRT_OPC_CTL_LS, &rpc);
+			D_ASSERTF(rc == 0,
+				   "crt_req_create failed; rc=%d\n", rc);
+
+			in_args = crt_req_get(rpc);
+			in_args->cel_grp_id = grp->cg_grpid;
+			in_args->cel_rank = rank;
+
+			rc = crt_req_set_timeout(rpc, ping_timeout);
+			D_ASSERTF(rc == 0,
+				   "crt_req_set_timeout failed; rc=%d\n", rc);
+
+			ws.rc = 0;
+			ws.num_ctx = 0;
+
+			rc = crt_req_send(rpc, ctl_client_cb, &ws);
+
+			if (rc == 0)
+				sync_timedwait(&ws, 120, __LINE__);
+
```

Comment:
same here, need else clause to set ws.rc = rc; otherwise we will exit loop on next iteration as ws.rc was set to 0 on line 165